### ISSUE TITLE
Building on Windows causes errors

### ISSUE THREAD

**Issue by hughbe** [dotnet/winforms#1230]
> git clone https://github.com/dotnet/winforms
> cd winforms
> build

```
Microsoft Windows [Version 10.0.17763.557]
(c) 2018 Microsoft Corporation. All rights reserved.

C:\Users\hughbe>git clone https://github.com/dotnet/winforms
Cloning into 'winforms'...
remote: Enumerating objects: 102, done.
remote: Counting objects: 100% (102/102), done.
remote: Compressing objects: 100% (75/75), done.
remote: Total 21689 (delta 42), reused 54 (delta 21), pack-reused 21587
Receiving objects: 100% (21689/21689), 23.34 MiB | 15.66 MiB/s, done.
Resolving deltas: 100% (16552/16552), done.
Checking out files: 100% (2362/2362), done.

C:\Users\hughbe>cd winforms

C:\Users\hughbe\winforms>build.cmd
dotnet-install: Downloading link: https://dotnetcli.azureedge.net/dotnet/Sdk/3.0.100-preview5-011568/dotnet-sdk-3.0.100-preview5-011568-win-x64.zip
dotnet-install: Extracting zip from https://dotnetcli.azureedge.net/dotnet/Sdk/3.0.100-preview5-011568/dotnet-sdk-3.0.100-preview5-011568-win-x64.zip
dotnet-install: Adding to current process PATH: "C:\Users\hughbe\winforms\.dotnet\". Note: This change will not be visible if PowerShell was run as a child process.
dotnet-install: Installation finished
Downloading vswhere
C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\MSBuild\Current\Bin\msbuild.exe /nologo /bl:C:\Users\hughbe\winforms\artifacts\log\Debug\ToolsetRestore.binlog /clp:Summary /clp:ErrorsOnly;NoSummary /m /nr:True /p:ContinuousIntegrationBuild=False /p:TreatWarningsAsErrors=true /p:__ToolsetLocationOutputFile=C:\Users\hughbe\winforms\artifacts\toolset\1.0.0-beta.19323.4.txt /t:__WriteToolsetLocation /v:minimal /warnaserror C:\Users\hughbe\winforms\artifacts\toolset\restore.proj
C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\MSBuild\Current\Bin\msbuild.exe /nologo /bl:C:\Users\hughbe\winforms\artifacts\log\Debug\Build.binlog /clp:Summary /m /nr:True /p:ContinuousIntegrationBuild=False /p:TreatWarningsAsErrors=true /p:Configuration=Debug /p:RepoRoot=C:\Users\hughbe\winforms /p:Restore=True /p:DeployDeps=False /p:Build=True /p:Rebuild=False /p:Deploy=False /p:Test=False /p:Pack=False /p:IntegrationTest=False /p:PerformanceTest=False /p:Sign=False /p:Publish=False /v:minimal /warnaserror C:\Users\hughbe\.nuget\packages\microsoft.dotnet.arcade.sdk\1.0.0-beta.19323.4\tools\Build.proj
C:\Users\hughbe\.nuget\packages\microsoft.dotnet.arcade.sdk\1.0.0-beta.19323.4\tools\Tools.proj : error : Unable to loc
ate the .NET Core SDK. Check that it is installed and that the version specified in global.json (if any) matches the in
stalled version.
C:\Users\hughbe\.nuget\packages\microsoft.dotnet.arcade.sdk\1.0.0-beta.19323.4\tools\Tools.proj : error MSB4236: The SD
K 'Microsoft.NET.Sdk' specified could not be found.

Build FAILED.

C:\Users\hughbe\.nuget\packages\microsoft.dotnet.arcade.sdk\1.0.0-beta.19323.4\tools\Tools.proj : error : Unable to loc
ate the .NET Core SDK. Check that it is installed and that the version specified in global.json (if any) matches the in
stalled version.
C:\Users\hughbe\.nuget\packages\microsoft.dotnet.arcade.sdk\1.0.0-beta.19323.4\tools\Tools.proj : error MSB4236: The SD
K 'Microsoft.NET.Sdk' specified could not be found.
    0 Warning(s)
    2 Error(s)

Time Elapsed 00:00:00.12
Build failed.
See log: C:\Users\hughbe\winforms\artifacts\log\Debug\Build.binlog

C:\Users\hughbe\winforms>
```

Any ideas?

**Comment by AdamYoblick** [comment]
I was not able to repro this locally, let me try again on a clean VM to make sure. I'll get back to you.

**Comment by AdamYoblick** [comment]
Out of curiosity, did you go through the setup required at https://github.com/dotnet/winforms/blob/master/Documentation/building.md ?

**Comment by hughbe** [comment]
Yes I ran build from root

**Comment by AdamYoblick** [comment]
Sorry, I meant the pre-requisites at https://github.com/dotnet/corefx/blob/master/Documentation/building/windows-instructions.md

**Comment by AdamYoblick** [comment]
I just created a clean win10 vm in azure and I followed your steps exactly. I was not able to repro even without installing any of the prerequisites.

I remember in the past I had issues creating/building repos in the c:\users\<username> folder. Can you please try from a folder outside of c:\users, maybe c:\src or something similar to see if your problem is resolved?

**Comment by RussKie** [comment]
Try running `PS> .\build-local -FullClean`

**Comment by vatsan-madhavan** [comment]
Do you really need the sdk entry in your global.json? If you removed it, wouldn't you just build using the tools (props/targets) from `tools.dotnet`, and build against ref-pack supplied by `runtimes.dotnet`?

**Comment by hughbe** [comment]
Can't repro anymore - I completely uninstalled .NET Core and reinstalled cleaning out all the paths etc which seemed to do the trick. Thanks all!

**Comment by RussKie** [comment]
For local build try `build-local.ps1` - that will download necessary sdks, if required, and symlink them.
You could also run `build-local.ps1 -FullClean` to delete existing sdks and setup afresh

**Comment by hughbe** [comment]
Running into problems again : (

> git clone https://github.com/dotnet/winforms
> cd winforms
> ./build-local.ps1 -FullClean

```
Windows PowerShell
Copyright (C) Microsoft Corporation. All rights reserved.

PS C:\Users\hughbe> git clone https://github.com/dotnet/winforms
Cloning into 'winforms'...
remote: Enumerating objects: 32, done.
remote: Counting objects: 100% (32/32), done.
remote: Compressing objects: 100% (23/23), done.
remote: Total 22981 (delta 20), reused 20 (delta 9), pack-reused 22949
Receiving objects: 100% (22981/22981), 24.59 MiB | 17.40 MiB/s, done.
Resolving deltas: 100% (17473/17473), done.
Checking out files: 100% (2479/2479), done.
PS C:\Users\hughbe> cd winforms
PS C:\Users\hughbe\winforms> .\build-local.ps1
Building the solution...
Processing C:\Users\hughbe\winforms\eng\..\global.json
File 'C:\Users\hughbe\.netcoreeng\native\temp\dotnet-api-docs_netcoreapp3.0-0.0.0.1-win64-x64.zip' already exists, skipping download
dotnet-api-docs_netcoreapp3.0 was not found in .
Microsoft.PowerShell.Commands.WriteErrorException: dotnet-api-docs_netcoreapp3.0 was not found in .
WARNING: dotnet-api-docs_netcoreapp3.0 installation failed
Processing C:\Users\hughbe\winforms\global.json
dotnet-api-docs_netcoreapp3.0 was not found in .
Microsoft.PowerShell.Commands.WriteErrorException: dotnet-api-docs_netcoreapp3.0 was not found in .
WARNING: dotnet-api-docs_netcoreapp3.0 installation failed
dotnet-install: Downloading link: https://dotnetcli.azureedge.net/dotnet/Sdk/3.0.100-preview5-011568/dotnet-sdk-3.0.100-preview5-011568-win-x64.zip
dotnet-install: Extracting zip from https://dotnetcli.azureedge.net/dotnet/Sdk/3.0.100-preview5-011568/dotnet-sdk-3.0.100-preview5-011568-win-x64.zip
dotnet-install: Adding to current process PATH: "C:\Users\hughbe\winforms\.dotnet\". Note: This change will not be visible if PowerShell was run as a child process.
dotnet-install: Installation finished
Downloading vswhere
C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\MSBuild\Current\Bin\msbuild.exe /nologo /bl:C:\Users\hughbe\winforms\artifacts\log\Debug\ToolsetRestore.binlog /clp:Summary /clp:ErrorsOnly;NoSummary /m /nr:True /p:ContinuousIntegrationBuild=False /p:TreatWarningsAsErrors=true /p:__ToolsetLocationOutputFile=C:\Users\hughbe\winforms\artifacts\toolset\1.0.0-beta.19323.4.txt /t:__WriteToolsetLocation /v:minimal /warnaserror C:\Users\hughbe\winforms\artifacts\toolset\restore.proj
C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\MSBuild\Current\Bin\msbuild.exe /nologo /bl:C:\Users\hughbe\winforms\artifacts\log\Debug\Build.binlog /clp:Summary /m /nr:True /p:ContinuousIntegrationBuild=False /p:TreatWarningsAsErrors=true /p:Configuration=Debug /p:RepoRoot=C:\Users\hughbe\winforms /p:Restore=True /p:DeployDeps=False /p:Build=True /p:Rebuild=False /p:Deploy=False /p:Test=False /p:Pack=False /p:IntegrationTest=False /p:PerformanceTest=False /p:Sign=False /p:Publish=False /v:minimal /warnaserror C:\Users\hughbe\.nuget\packages\microsoft.dotnet.arcade.sdk\1.0.0-beta.19323.4\tools\Build.proj
C:\Users\hughbe\.nuget\packages\microsoft.dotnet.arcade.sdk\1.0.0-beta.19323.4\tools\Tools.proj : error : Unable to loc
ate the .NET Core SDK. Check that it is installed and that the version specified in global.json (if any) matches the in
stalled version.
C:\Users\hughbe\.nuget\packages\microsoft.dotnet.arcade.sdk\1.0.0-beta.19323.4\tools\Tools.proj : error MSB4236: The SD
K 'Microsoft.NET.Sdk' specified could not be found.

Build FAILED.

C:\Users\hughbe\.nuget\packages\microsoft.dotnet.arcade.sdk\1.0.0-beta.19323.4\tools\Tools.proj : error : Unable to loc
ate the .NET Core SDK. Check that it is installed and that the version specified in global.json (if any) matches the in
stalled version.
C:\Users\hughbe\.nuget\packages\microsoft.dotnet.arcade.sdk\1.0.0-beta.19323.4\tools\Tools.proj : error MSB4236: The SD
K 'Microsoft.NET.Sdk' specified could not be found.
    0 Warning(s)
    2 Error(s)

Time Elapsed 00:00:00.17
Build failed.
See log: C:\Users\hughbe\winforms\artifacts\log\Debug\Build.binlog
√ Solution built
Creating a symlink...
System.ComponentModel.Win32Exception (0x80004005): The system cannot find the path specified
   at Microsoft.PowerShell.Commands.FileSystemProvider.NewItem(String path, String type, Object value)
   at System.Management.Automation.SessionStateInternal.NewItemPrivate(CmdletProvider providerInstance, String path, String type, Object content, CmdletProviderContext context)
Press Enter to continue...:
PS C:\Users\hughbe\winforms>
```

**Comment by hughbe** [comment]
I was under the impression that I don't need a globally install `dotnet` to make this work

**Comment by RussKie** [comment]
> System.ComponentModel.Win32Exception (0x80004005): The system cannot find the path specified

This is an odd one.
Are you running the script on a brand new machine?

try adding a `Write-Host` statement before this line and see what path it is trying to link:
https://github.com/dotnet/winforms/blob/master/build-local.ps1#L96

**Comment by vatsan-madhavan** [comment]
@RussKie, consider creating an instrumented branch that @hughbe can checkout and build. 

I tried master under Process Explorer and didn’t see any $env:ProgramFiles usage for the SDK. The only oddness I noticed which I wasn’t happy about was a junction point being created _from_ $env:ProgramFile\dotnet\... to $(RepoRoot).dotnet\sdk\...  at the end of the build.

**Comment by jdm7dv** [comment]
They are treating warnings are errors in the microsoft.dotnet.arcade.sdk\1.0.0-beta.19323.4

**Comment by AdamYoblick** [comment]
> You really shouldn't depend on the internet to build software.

This is kind of off-topic, but I'm not sure I understand this comment. The whole point of a package manager like nuget is specifically so you don't have to check packages into source control. It is literally depending on the internet to build software.

Igor, this seems like maybe a bug with the build-local script if the user has never installed any dotnet sdk?

**Comment by jdm7dv** [comment]
![Untitled](https://user-images.githubusercontent.com/8218836/61335819-dc3b9400-a7fc-11e9-851d-0bd63205c37c.png)

Might not be the problem but it was a thought. You can build C# or .NET projects with a simple makefile and nmake might not resolve all of the dependencies but in the old .NET we didn't need many.

**Comment by jdm7dv** [comment]
[https://github.com/intj-t/Lex](url) Here is an example in C# with nmake if anyone cares.

**Comment by RussKie** [comment]
> Igor, this seems like maybe a bug with the build-local script if the user has never installed any dotnet sdk?

I'm going to test it on a blank VM

**Comment by AdamYoblick** [comment]
@hughbe Is this still an issue?
